# ***My answers for Problem 2 are at the bottom, after the template notebook code**

# Machine Learning Exercise 4 - Neural Networks

This notebook covers a Python-based solution for the fourth programming exercise of the machine learning class on Coursera.  Please refer to the [exercise text](https://github.com/jdwittenauer/ipython-notebooks/blob/master/exercises/ML/ex4.pdf) for detailed descriptions and equations.

For this exercise we'll again tackle the hand-written digits data set, this time using a feed-forward neural network with backpropagation.  We'll implement un-regularized and regularized versions of the neural network cost function and gradient computation via the backpropagation algorithm.  We'll also implement random weight initialization and a method to use the network to make predictions.

Since the data set is the same one we used in exercise 3, we'll re-use the code to load the data.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.io import loadmat
%matplotlib inline

data = loadmat('ex3data1.mat')
data

{'__header__': b'MATLAB 5.0 MAT-file, Platform: GLNXA64, Created on: Sun Oct 16 13:09:09 2011',
 '__version__': '1.0',
 '__globals__': [],
 'X': array([[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]]),
 'y': array([[10],
        [10],
        [10],
        ...,
        [ 9],
        [ 9],
        [ 9]], dtype=uint8)}

Since we're going to need these later (and will use them often), let's create some useful variables up-front.

In [2]:
X = data['X']
y = data['y']

X.shape, y.shape

((5000, 400), (5000, 1))

We're also going to need to one-hot encode our y labels.  One-hot encoding turns a class label n (out of k classes) into a vector of length k where index n is "hot" (1) while the rest are zero.  Scikit-learn has a built in utility we can use for this.

In [3]:
from sklearn.preprocessing import OneHotEncoder
encoder = OneHotEncoder(sparse_output=False)
y_onehot = encoder.fit_transform(y)
y_onehot.shape

(5000, 10)

In [4]:
y[0], y_onehot[0,:]

(array([10], dtype=uint8), array([0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]))

The neural network we're going to build for this exercise has an input layer matching the size of our instance data (400 + the bias unit), a hidden layer with 25 units (26 with the bias unit), and an output layer with 10 units corresponding to our one-hot encoding for the class labels.  For additional details and an image of the network architecture, please refer to the PDF in the "exercises" folder.

The first piece we need to implement is a cost function to evaluate the loss for a given set of network parameters.  The source mathematical function is in the exercise text (and looks pretty intimidating).  Here are the functions required to compute the cost.

In [5]:
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

In [6]:
def forward_propagate(X, theta1, theta2):
    m = X.shape[0]

    a1 = np.insert(X, 0, values=np.ones(m), axis=1)
    z2 = a1 * theta1.T
    a2 = np.insert(sigmoid(z2), 0, values=np.ones(m), axis=1)
    z3 = a2 * theta2.T
    h = sigmoid(z3)

    return a1, z2, a2, z3, h

In [7]:
def cost(params, input_size, hidden_size, num_labels, X, y, learning_rate):
    m = X.shape[0]
    X = np.matrix(X)
    y = np.matrix(y)

    # reshape the parameter array into parameter matrices for each layer
    theta1 = np.matrix(np.reshape(params[:hidden_size * (input_size + 1)], (hidden_size, (input_size + 1))))
    theta2 = np.matrix(np.reshape(params[hidden_size * (input_size + 1):], (num_labels, (hidden_size + 1))))

    # run the feed-forward pass
    a1, z2, a2, z3, h = forward_propagate(X, theta1, theta2)

    # compute the cost
    J = 0
    for i in range(m):
        first_term = np.multiply(-y[i,:], np.log(h[i,:]))
        second_term = np.multiply((1 - y[i,:]), np.log(1 - h[i,:]))
        J += np.sum(first_term - second_term)

    J = J / m

    return J

We've used the sigmoid function before so that's not new.  The forward-propagate function computes the hypothesis for each training instance given the current parameters.  It's output shape should match the same of our one-hot encoding for y.  We can test this real quick to convince ourselves that it's working as expected (the intermediate steps are also returned as these will be useful later).

In [8]:
# initial setup
input_size = 400
hidden_size = 25
num_labels = 10
learning_rate = 1

# randomly initialize a parameter array of the size of the full network's parameters
params = (np.random.random(size=hidden_size * (input_size + 1) + num_labels * (hidden_size + 1)) - 0.5) * 0.25

m = X.shape[0]
X = np.matrix(X)
y = np.matrix(y)

# unravel the parameter array into parameter matrices for each layer
theta1 = np.matrix(np.reshape(params[:hidden_size * (input_size + 1)], (hidden_size, (input_size + 1))))
theta2 = np.matrix(np.reshape(params[hidden_size * (input_size + 1):], (num_labels, (hidden_size + 1))))

theta1.shape, theta2.shape

((25, 401), (10, 26))

In [9]:
a1, z2, a2, z3, h = forward_propagate(X, theta1, theta2)
a1.shape, z2.shape, a2.shape, z3.shape, h.shape

((5000, 401), (5000, 25), (5000, 26), (5000, 10), (5000, 10))

The cost function, after computing the hypothesis matrix h, applies the cost equation to compute the total error between y and h.

In [10]:
cost(params, input_size, hidden_size, num_labels, X, y_onehot, learning_rate)

np.float64(7.197876556420712)

Our next step is to add regularization to the cost function.  If you're following along in the exercise text and thought the last equation looked ugly, this one looks REALLY ugly.  It's actually not as complicated as it looks though - in fact, the regularization term is simply an addition to the cost we already computed.  Here's the revised cost function.

In [11]:
def cost(params, input_size, hidden_size, num_labels, X, y, learning_rate):
    m = X.shape[0]
    X = np.matrix(X)
    y = np.matrix(y)

    # reshape the parameter array into parameter matrices for each layer
    theta1 = np.matrix(np.reshape(params[:hidden_size * (input_size + 1)], (hidden_size, (input_size + 1))))
    theta2 = np.matrix(np.reshape(params[hidden_size * (input_size + 1):], (num_labels, (hidden_size + 1))))

    # run the feed-forward pass
    a1, z2, a2, z3, h = forward_propagate(X, theta1, theta2)

    # compute the cost
    J = 0
    for i in range(m):
        first_term = np.multiply(-y[i,:], np.log(h[i,:]))
        second_term = np.multiply((1 - y[i,:]), np.log(1 - h[i,:]))
        J += np.sum(first_term - second_term)

    J = J / m

    # add the cost regularization term
    J += (float(learning_rate) / (2 * m)) * (np.sum(np.power(theta1[:,1:], 2)) + np.sum(np.power(theta2[:,1:], 2)))

    return J

In [12]:
cost(params, input_size, hidden_size, num_labels, X, y_onehot, learning_rate)

np.float64(7.203279011957698)

Next up is the backpropagation algorithm.  Backpropagation computes the parameter updates that will reduce the error of the network on the training data.  The first thing we need is a function that computes the gradient of the sigmoid function we created earlier.

In [13]:
def sigmoid_gradient(z):
    return np.multiply(sigmoid(z), (1 - sigmoid(z)))

Now we're ready to implement backpropagation to compute the gradients.  Since the computations required for backpropagation are a superset of those required in the cost function, we're actually going to extend the cost function to also perform backpropagation and return both the cost and the gradients.

In [14]:
def backprop(params, input_size, hidden_size, num_labels, X, y, learning_rate):
    m = X.shape[0]
    X = np.matrix(X)
    y = np.matrix(y)

    # reshape the parameter array into parameter matrices for each layer
    theta1 = np.matrix(np.reshape(params[:hidden_size * (input_size + 1)], (hidden_size, (input_size + 1))))
    theta2 = np.matrix(np.reshape(params[hidden_size * (input_size + 1):], (num_labels, (hidden_size + 1))))

    # run the feed-forward pass
    a1, z2, a2, z3, h = forward_propagate(X, theta1, theta2)

    # initializations
    J = 0
    delta1 = np.zeros(theta1.shape)  # (25, 401)
    delta2 = np.zeros(theta2.shape)  # (10, 26)

    # compute the cost
    for i in range(m):
        first_term = np.multiply(-y[i,:], np.log(h[i,:]))
        second_term = np.multiply((1 - y[i,:]), np.log(1 - h[i,:]))
        J += np.sum(first_term - second_term)

    J = J / m

    # add the cost regularization term
    J += (float(learning_rate) / (2 * m)) * (np.sum(np.power(theta1[:,1:], 2)) + np.sum(np.power(theta2[:,1:], 2)))

    # perform backpropagation
    for t in range(m):
        a1t = a1[t,:]  # (1, 401)
        z2t = z2[t,:]  # (1, 25)
        a2t = a2[t,:]  # (1, 26)
        ht = h[t,:]  # (1, 10)
        yt = y[t,:]  # (1, 10)

        d3t = ht - yt  # (1, 10)

        z2t = np.insert(z2t, 0, values=np.ones(1))  # (1, 26)
        d2t = np.multiply((theta2.T * d3t.T).T, sigmoid_gradient(z2t))  # (1, 26)

        delta1 = delta1 + (d2t[:,1:]).T * a1t
        delta2 = delta2 + d3t.T * a2t

    delta1 = delta1 / m
    delta2 = delta2 / m

    # unravel the gradient matrices into a single array
    grad = np.concatenate((np.ravel(delta1), np.ravel(delta2)))

    return J, grad

The hardest part of the backprop computation (other than understanding WHY we're doing all these calculations) is getting the matrix dimensions right.  By the way, if you find it confusing when to use A * B vs. np.multiply(A, B), you're not alone.  Basically the former is a matrix multiplication and the latter is an element-wise multiplication (unless A or B is a scalar value, in which case it doesn't matter).  I wish there was a more concise syntax for this (maybe there is and I'm just not aware of it).

Anyway, let's test it out to make sure the function returns what we're expecting it to.

In [15]:
J, grad = backprop(params, input_size, hidden_size, num_labels, X, y_onehot, learning_rate)
J, grad.shape

(np.float64(7.203279011957698), (10285,))

We still have one more modification to make to the backprop function - adding regularization to the gradient calculations.  The final regularized version is below.

In [16]:
def backprop(params, input_size, hidden_size, num_labels, X, y, learning_rate):
    m = X.shape[0]
    X = np.matrix(X)
    y = np.matrix(y)

    # reshape the parameter array into parameter matrices for each layer
    theta1 = np.matrix(np.reshape(params[:hidden_size * (input_size + 1)], (hidden_size, (input_size + 1))))
    theta2 = np.matrix(np.reshape(params[hidden_size * (input_size + 1):], (num_labels, (hidden_size + 1))))

    # run the feed-forward pass
    a1, z2, a2, z3, h = forward_propagate(X, theta1, theta2)

    # initializations
    J = 0
    delta1 = np.zeros(theta1.shape)  # (25, 401)
    delta2 = np.zeros(theta2.shape)  # (10, 26)

    # compute the cost
    for i in range(m):
        first_term = np.multiply(-y[i,:], np.log(h[i,:]))
        second_term = np.multiply((1 - y[i,:]), np.log(1 - h[i,:]))
        J += np.sum(first_term - second_term)

    J = J / m

    # add the cost regularization term
    J += (float(learning_rate) / (2 * m)) * (np.sum(np.power(theta1[:,1:], 2)) + np.sum(np.power(theta2[:,1:], 2)))

    # perform backpropagation
    for t in range(m):
        a1t = a1[t,:]  # (1, 401)
        z2t = z2[t,:]  # (1, 25)
        a2t = a2[t,:]  # (1, 26)
        ht = h[t,:]  # (1, 10)
        yt = y[t,:]  # (1, 10)

        d3t = ht - yt  # (1, 10)

        z2t = np.insert(z2t, 0, values=np.ones(1))  # (1, 26)
        d2t = np.multiply((theta2.T * d3t.T).T, sigmoid_gradient(z2t))  # (1, 26)

        delta1 = delta1 + (d2t[:,1:]).T * a1t
        delta2 = delta2 + d3t.T * a2t

    delta1 = delta1 / m
    delta2 = delta2 / m

    # add the gradient regularization term
    delta1[:,1:] = delta1[:,1:] + (theta1[:,1:] * learning_rate) / m
    delta2[:,1:] = delta2[:,1:] + (theta2[:,1:] * learning_rate) / m

    # unravel the gradient matrices into a single array
    grad = np.concatenate((np.ravel(delta1), np.ravel(delta2)))

    return J, grad

In [17]:
J, grad = backprop(params, input_size, hidden_size, num_labels, X, y_onehot, learning_rate)
J, grad.shape

(np.float64(7.203279011957698), (10285,))

We're finally ready to train our network and use it to make predictions.  This is roughly similar to the previous exercise with multi-class logistic regression.

In [18]:
from scipy.optimize import minimize

options={'maxfun': 250}

# minimize the objective function
fmin = minimize(fun=backprop, x0=params, args=(input_size, hidden_size, num_labels, X, y_onehot, learning_rate),
                method='TNC', jac=True, options=options)
fmin

 message: Max. number of function evaluations reached
 success: False
  status: 3
     fun: 0.33780922572841104
       x: [ 2.150e+00 -8.265e-03 ... -1.084e+00 -1.043e+00]
     nit: 22
     jac: [ 7.247e-07 -1.653e-06 ... -2.479e-04 -5.603e-05]
    nfev: 250

We put a bound on the number of iterations since the objective function is not likely to completely converge.  Our total cost has dropped below 0.5 though so that's a good indicator that the algorithm is working.  Let's use the parameters it found and forward-propagate them through the network to get some predictions.

In [19]:
X = np.matrix(X)
theta1 = np.matrix(np.reshape(fmin.x[:hidden_size * (input_size + 1)], (hidden_size, (input_size + 1))))
theta2 = np.matrix(np.reshape(fmin.x[hidden_size * (input_size + 1):], (num_labels, (hidden_size + 1))))

a1, z2, a2, z3, h = forward_propagate(X, theta1, theta2)
y_pred = np.array(np.argmax(h, axis=1) + 1)
y_pred

array([[10],
       [10],
       [10],
       ...,
       [ 9],
       [ 9],
       [ 9]])

Finally we can compute the accuracy to see how well our trained network is doing.

In [20]:
correct = [1 if a == b else 0 for (a, b) in zip(y_pred, y)]
accuracy = (sum(map(int, correct)) / float(len(correct)))
print ('accuracy = {0}%'.format(accuracy * 100))

accuracy = 99.22%


And we're done!  We've successfully implemented a rudimentary feed-forward neural network with backpropagation and used it to classify images of handwritten digits.  In the next exercise we'll look at another power supervised learning algorithm, support vector machines.

# **My answers to Problem 2 start here:**

In [21]:
X = data['X']
y = y.flatten()

y_adj = y.copy()
y_adj[y_adj == 10] = 0

X.shape, y.shape

((5000, 400), (1, 5000))

In [22]:
y_adj = np.array(y_adj).flatten()
encoder = OneHotEncoder(sparse_output=False)
y_encoded = encoder.fit_transform(y_adj.reshape(-1, 1))

### 1. Scaled Sigmoid

In [23]:
def scaled_sigmoid(z):
    return 1 / (1 + np.exp(-2 * z))

### 4. Sigmoid Gradient

In [24]:
def scaled_sigmoid_grad(z):
    s = scaled_sigmoid(z)
    return 2 * s * (1 - s)

In [25]:
def init_weights_rand(fan_in, fan_out, eps_init=0.12):
    return np.random.rand(fan_out, fan_in + 1) * 2 * eps_init - eps_init

def updated_init_params(input_size, hidden_size1, hidden_size2, num_labels):
    theta1 = init_weights_rand(input_size, hidden_size1)
    theta2 = init_weights_rand(hidden_size1, hidden_size2)
    theta3 = init_weights_rand(hidden_size2, num_labels)
    return np.concatenate([theta1.ravel(), theta2.ravel(), theta3.ravel()])

In [26]:
def updated_unpack_params(params, input_size, hidden_size1, hidden_size2, num_labels):
    end1 = hidden_size1 * (input_size + 1)
    end2 = end1 + hidden_size2 * (hidden_size1 + 1)

    theta1 = np.reshape(params[:end1], (hidden_size1, input_size + 1))
    theta2 = np.reshape(params[end1:end2], (hidden_size2, hidden_size1 + 1))
    theta3 = np.reshape(params[end2:], (num_labels, hidden_size2 + 1))
    return theta1, theta2, theta3

### 2. Forward Propagation

In [27]:
def updated_forward_propagate(X, theta1, theta2, theta3):
    m = X.shape[0]

    a1 = np.insert(X, 0, 1, axis=1)
    z2 = a1 @ theta1.T
    a2 = scaled_sigmoid(z2)

    a2 = np.insert(a2, 0, 1, axis=1)
    z3 = a2 @ theta2.T
    a3 = scaled_sigmoid(z3)

    a3 = np.insert(a3, 0, 1, axis=1)
    z4 = a3 @ theta3.T
    a4 = scaled_sigmoid(z4)

    return a1, z2, a2, z3, a3, z4, a4

### 3. Cost Function

In [28]:
def updated_cost(params, input_size, hidden_size1, hidden_size2, num_labels, X, y_encoded, reg_strength=0.0):
    m = X.shape[0]

    end1 = hidden_size1 * (input_size + 1)
    end2 = end1 + hidden_size2 * (hidden_size1 + 1)

    theta1 = np.reshape(params[:end1], (hidden_size1, input_size + 1))
    theta2 = np.reshape(params[end1:end2], (hidden_size2, hidden_size1 + 1))
    theta3 = np.reshape(params[end2:], (num_labels, hidden_size2 + 1))

    _, _, _, _, _, _, h = updated_forward_propagate(X, theta1, theta2, theta3)

    term1 = -y_encoded * np.log(h)
    term2 = (1 - y_encoded) * np.log(1 - h)
    J = np.sum(term1 - term2) / m

    # added this function so it works with and without regularization
    if reg_strength > 0:
        reg = (np.sum(np.square(theta1[:, 1:])) +
               np.sum(np.square(theta2[:, 1:])) +
               np.sum(np.square(theta3[:, 1:])))
        J += (reg_strength / (2 * m)) * reg

    return J


### 5. Backpropagation

In [29]:
def updated_backprop(params, input_size, hidden_size1, hidden_size2, num_labels, X, y_encoded, reg_strength=0.0):
    m = X.shape[0]

    end1 = hidden_size1 * (input_size + 1)
    end2 = end1 + hidden_size2 * (hidden_size1 + 1)

    theta1 = np.reshape(params[:end1], (hidden_size1, input_size + 1))
    theta2 = np.reshape(params[end1:end2], (hidden_size2, hidden_size1 + 1))
    theta3 = np.reshape(params[end2:], (num_labels, hidden_size2 + 1))

    a1, z2, a2, z3, a3, z4, h = updated_forward_propagate(X, theta1, theta2, theta3)

    delta4 = h - y_encoded
    delta3 = delta4 @ theta3[:, 1:] * scaled_sigmoid_grad(z3)
    delta2 = delta3 @ theta2[:, 1:] * scaled_sigmoid_grad(z2)

    d1 = delta2.T @ a1
    d2 = delta3.T @ a2
    d3 = delta4.T @ a3

    if reg_strength > 0:
        d1[:, 1:] += reg_strength * theta1[:, 1:]
        d2[:, 1:] += reg_strength * theta2[:, 1:]
        d3[:, 1:] += reg_strength * theta3[:, 1:]

    d1 /= m
    d2 /= m
    d3 /= m

    grad = np.concatenate([d1.ravel(), d2.ravel(), d3.ravel()])
    J = updated_cost(params, input_size, hidden_size1, hidden_size2, num_labels, X, y_encoded, reg_strength)
    return J, grad

In [33]:
def predict_3layer(X, theta1, theta2, theta3):
    _, _, _, _, _, _, h = updated_forward_propagate(X, theta1, theta2, theta3)
    return np.argmax(h, axis=1)

### 6. Training 3-Layered Neural Network

In [34]:
from scipy.optimize import minimize

# training
input_size = 400
hidden_size1 = 20
hidden_size2 = 20
num_labels = 10

def init_weights_rand(fan_in, fan_out, eps_init=0.12):
    return np.random.rand(fan_out, fan_in + 1) * 2 * eps_init - eps_init

init_params = np.concatenate([
    init_weights_rand(input_size, hidden_size1).ravel(),
    init_weights_rand(hidden_size1, hidden_size2).ravel(),
    init_weights_rand(hidden_size2, num_labels).ravel()
])

options = {'maxiter': 250}
reg_strength = 1

result = minimize(fun=updated_backprop,
                  x0=init_params,
                  args=(input_size, hidden_size1, hidden_size2, num_labels, X, y_encoded, reg_strength),
                  method='TNC',
                  jac=True,
                  options=options)

<ipython-input-34-5c89e84f32ff>:21: OptimizeWarning: Unknown solver options: maxiter
  result = minimize(fun=updated_backprop,


### 7. Making Forward Predictions From the Trained Model & Computing Accuracy

In [35]:
# unrolling the learned parameters
end1 = hidden_size1 * (input_size + 1)
end2 = end1 + hidden_size2 * (hidden_size1 + 1)

theta1 = np.reshape(result.x[:end1], (hidden_size1, input_size + 1))
theta2 = np.reshape(result.x[end1:end2], (hidden_size2, hidden_size1 + 1))
theta3 = np.reshape(result.x[end2:], (num_labels, hidden_size2 + 1))

# computing the accuracy
y_pred = predict_3layer(X, theta1, theta2, theta3)
accuracy = np.mean(y_pred == y_adj) * 100
print(f"Accuracy of 3-layer Neural Network: {accuracy:.2f}%")

Accuracy of 3-layer Neural Network: 100.00%


### 8. Observations

When comparing the performance of the two models, the 3-layer neural network achieved a slightly higher accuracy than the 2-layer network. The 2-layer model reached around 99.22% accuracy, while the 3-layer model achieved 100% accuracy on the training data. This small improvement in training accuracy suggests that the deeper network was able to fit the data more precisely, which makes sense since adding another hidden layer increases the model's capacity to learn more complex patterns. However, getting perfect accuracy on training data may also be a sign of overfitting rather than better generalization. Since we are only evaluating on the training set, it's hard to tell if the 3-layer network would actually perform better on new, unseen data. Still, under the current setup, the deeper network does outperform the simpler one in terms of training accuracy.